src/persist-txn/src/txn_write.rs

```diff
@@ -179,62 +211,302 @@ where
                     for batch in txn_batches {
                         let () = batch.delete().await;
                     }
+                    let () = handle.txns_cache.update_ge(&txns_upper).await;
                     continue;
                 }
             }
         }
     }
+
+    /// Merges the staged writes in the other txn into this one.
+    pub fn merge(&mut self, other: Self) {
+        for (data_id, writes) in other.writes {
+            self.writes.entry(data_id).or_default().extend(writes);
+        }
+        self.tidy.merge(other.tidy);
+    }
+
+    /// Merges the work represented by given tidy into this txn.
+    ///
+    /// If this txn commits, the tidy work will be written at the commit ts.
+    pub fn tidy(&mut self, tidy: Tidy) {
+        self.tidy.merge(tidy);
+    }
+
+    /// Extracts any tidy work that has been merged into this txn with
+    /// [Self::tidy].
+    pub fn take_tidy(&mut self) -> Tidy {
+        std::mem::take(&mut self.tidy)
+    }
 }
 
 /// A token representing the asynchronous "apply" work expected to be promptly
 /// performed by a txn committer.
 #[derive(Debug)]
+#[cfg_attr(any(test, debug_assertions), derive(PartialEq))]
 pub struct TxnApply<T> {
+    is_empty: bool,
     pub(crate) commit_ts: T,
 }
 
 impl<T> TxnApply<T> {
     /// Applies the txn, unblocking reads at timestamp it was committed at.
-    pub async fn apply<K, V, D>(self, handle: &mut TxnsHandle<K, V, T, D>)
+    pub async fn apply<K, V, D, C>(self, handle: &mut TxnsHandle<K, V, T, D, C>) -> Tidy
```

Comment:
So the typical usage would look something like this? (assuming no conflicts):

```
let txns: TxnHandle<_> = ...;

// register some shards.

let mut txn = txns.begin();

// write some data.

let ts: T = ...;
let tidy = txn.commit_at(&mut txns, ts).await.unwrap().apply(&mut txns).await;
txns.read_cache().update_ge(ts + 1).await;
txns.tidy_at(ts + 1, tidy).unwrap();
```

It might be useful to update the example usages in `lib.rs` with this (up to you).

Reply:
Done! I think in practice it'll be more common to just sneak the tidy into the next commit, so modeled that in the doctest (also no need for the crate user to do the update_ge, the commit process will do that for us)